<a href="https://colab.research.google.com/github/Rochy3/AlgorithmsUN2024II/blob/main/Lab3/RosaSantofimio_Lab3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
!cat /proc/cpuinfo

processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 79
model name	: Intel(R) Xeon(R) CPU @ 2.20GHz
stepping	: 0
microcode	: 0xffffffff
cpu MHz		: 2199.998
cache size	: 56320 KB
physical id	: 0
siblings	: 2
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm rdseed adx smap xsaveopt arat md_clear arch_capabilities
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs taa mmio_stale_data retbleed bhi
bogomips	: 4399.99
clflush size	: 64
cache_alignment	: 64
ad

# I Solving the Sum of Two Digits Programming Challenge

## C++

In [4]:
%%writefile APlusB.cpp
#include <iostream>

int sum_of_two_digits(int first_digit, int second_digit) {
    return first_digit + second_digit;
}

int main() {
    int a = 0;
    int b = 0;
    std::cin >> a;
    std::cin >> b;
    std::cout << sum_of_two_digits(a, b);
    return 0;
}

Writing APlusB.cpp


In [5]:
!g++ APlusB.cpp -o APlusB -std=c++11
!./APlusB

2
1
3

### Test input output files

In [6]:
%%writefile input1.txt
4 8

Writing input1.txt


In [7]:
%%writefile output1.txt
12

Writing output1.txt


In [8]:
%%writefile input2.txt
0 9

Writing input2.txt


In [9]:
%%writefile output2.txt
9

Writing output2.txt


In [10]:
!./APlusB < input1.txt

12

In [11]:
!./APlusB < input2.txt

9

### Comparing files with grep

https://askubuntu.com/questions/546796/comparing-two-text-files

In [12]:
!grep -xvFf output1.txt output1.txt

In [13]:
!grep -xvFf output1.txt output2.txt

9


In [14]:
!grep -xvFf output2.txt output1.txt

12


In [15]:
!./APlusB < input1.txt> ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [16]:
!./APlusB < input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

## Python

In [17]:
%%writefile sum.py
# python3

def sum_of_two_digits(first_digit, second_digit):
    return first_digit + second_digit

if __name__ == '__main__':
    a, b = map(int, input().split())
    print(sum_of_two_digits(a, b))

Writing sum.py


escribir los siguientes digitos en la misma linea

In [158]:
!python sum.py

4 5
9


In [19]:
!python sum.py  < input1.txt> ptest1.txt
!grep -xvFf output1.txt ptest1.txt

In [20]:
!python sum.py  < input2.txt> ptest2.txt
!grep -xvFf output2.txt ptest2.txt

# II Solving The Maximum Pairwise Product Programming Challenge in C++

## C++

$$ T(n) < 6 n^2 = O(n^2)$$

In [21]:
%%writefile max_pairwise_product_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int aux_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second ){
                aux_product = numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product_full.cpp


#### Time Complexity
$$\left. \begin{array}{llc}
 \text{INSTRUCTION} & \text{TYPE}& \text{# EXECUTIONS}\\
\text{first != second} &  \text{comparisons} &n^2 \\
\text{numbers[first] * numbers[second]}& \ \text{ products}  &n^2 - n \\
\text{max_product = aux_product} &  \text{assigments} & [1, \frac{n^2 - n}{2})  
\end{array}\right\}  =  O(n^2)$$

In [22]:
!g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
!./max_pairwise_product_full

1
2
0


### Test 1: Small numbers

In [23]:
%%writefile input1.txt
10
1 2 3 4 5 6 7 8 9 10

Overwriting input1.txt


In [24]:
!./max_pairwise_product < input1.txt

/bin/bash: line 1: ./max_pairwise_product: No such file or directory


In [25]:
%%writefile output1.txt
90

Overwriting output1.txt


In [26]:
!./max_pairwise_product_full< input1.txt > ctest1.txt
!grep -xvFf output1.txt ctest1.txt

In [27]:
%%writefile input2.txt
10
10 9 8 7 6 5 4 3 2 1

Overwriting input2.txt


In [28]:
%%writefile output2.txt
90

Overwriting output2.txt


In [29]:
!./max_pairwise_product_full< input2.txt > ctest2.txt
!grep -xvFf output2.txt ctest2.txt

### Test 2: largest input numbers 200000 200000 to check possible intger overflow

In [30]:
%%writefile input3.txt
6
1 3 200000 9 200000 23

Writing input3.txt


In [31]:
200000*200000

40000000000

In [32]:
%%writefile output3.txt
40000000000

Writing output3.txt


In [33]:
!./max_pairwise_product_full < input3.txt

1345294336


In [34]:
!./max_pairwise_product_full < input3.txt > ctest3.txt
!grep -xvFf output3.txt ctest3.txt

1345294336


In [35]:
%%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_full.cpp -o max_pairwise_product_full
for i in {0..3}
do
    ./max_pairwise_product_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Writing run_all_test.sh


In [191]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

1345294336


### Changing the variables to long long

In [37]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;
	  long long aux_product = 0LL;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = 0; second < n; ++second) {
          if ( first != second){
                aux_product = (long long) numbers[first] * numbers[second];
                if (aux_product > max_product)  max_product = aux_product;
          }
        }
    }


    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Writing max_pairwise_product_long_full.cpp


In [38]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full

In [39]:
!./max_pairwise_product_long_full < input3.txt

40000000000


### Runing multuple test with a bash shell

In [40]:
 %%writefile run_all_test.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
for i in {1..3}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    grep -xvFf output$i.txt ctest$i.txt
done

Overwriting run_all_test.sh


In [41]:
!chmod u+x run_all_test.sh
!./run_all_test.sh

### Test 3  stress tests, 6 testing cases files input4.txt/output4.txt .. input10.txt/output10.txt

---



In [42]:

%%writefile input4.txt
7
1 3 200000 9 200000 23 200000

Writing input4.txt


In [43]:

%%writefile output4.txt
40000000000

Writing output4.txt


In [44]:

%%writefile input5.txt
2
1 1

Writing input5.txt


In [45]:

%%writefile output5.txt
1

Writing output5.txt


In [46]:
%%writefile input6.txt
8
2 2 2 2 2 2 2 2

Writing input6.txt


In [47]:
%%writefile output6.txt
4

Writing output6.txt


In [48]:
%%writefile input7.txt
5
1 3 200000 9 200000


Writing input7.txt


In [49]:
%%writefile output7.txt
40000000000

Writing output7.txt


In [50]:
%%writefile input8.txt
3
9 8 9

Writing input8.txt


In [51]:
%%writefile output8.txt
81

Writing output8.txt


In [52]:

%%writefile input9.txt
4
200000 200000 200000 200000

Writing input9.txt


In [53]:
%%writefile output9.txt
40000000000

Writing output9.txt


In [54]:
%%writefile input10.txt
3
0 0 0

Writing input10.txt


In [55]:

%%writefile output10.txt
0


Writing output10.txt


In [56]:
%%writefile run_all_test_1.sh
#!/bin/bash
g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full || exit 1

for i in {4..10}
do
    ./max_pairwise_product_long_full < input$i.txt > ctest$i.txt
    if grep -xvFf output$i.txt ctest$i.txt; then
        echo "Test $i failed!"
    else
        echo "Test $i passed!"
    fi
done


Writing run_all_test_1.sh


In [57]:
!chmod u+x run_all_test_1.sh
!./run_all_test_1.sh

Test 4 passed!
Test 5 passed!
Test 6 passed!
Test 7 passed!
Test 8 passed!
Test 9 passed!
Test 10 passed!


### Test 4  time test  measure the execution time running with the the largest worst cases input11.txt  (200000 increasing numbers) and input12.txt (199998 increasing numbers and  200000, 200000 at the end)

In [58]:
# input11.txt 200000 increasing numbers:  1 2 .. 200000

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input11.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Writing WriteTest.cpp


In [59]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [60]:
199999 * 200000

39999800000

In [61]:
%%writefile output11.txt
39999800000

Writing output11.txt


In [62]:
# input12.txt 200000 decreasing numbers and 200000, 199999 ...  2 1

%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input12.txt");

 MyFile << 200000 <<"\n";  // n
 for (int i=1; i <200001; i++){
     MyFile << 200001-i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [63]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [64]:
%%writefile output12.txt
39999800000

Writing output12.txt


### Measure time

https://askubuntu.com/questions/53444/how-can-i-measure-the-execution-time-of-a-terminal-process

In [192]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full < input11.txt

123243423
^C


In [193]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input12.txt

123213 2312
^C


### Calculate the number or and upper bound to the number of instructions executed by the algorithm as a function of n, add in each line a comment with the number of instructions that are executed in that line.

In [67]:
%%writefile max_pairwise_product_long_full.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = 0; second < n; ++second) {                         //2n + 2n**2
          if ( first != second ){                                            //n**2
                aux_product = numbers[first] * numbers[second];              //n**2 -n
                if (aux_product > max_product)                               //n**2 -n
                    max_product = aux_product;                               //[0, n**2/2 - n/2] upper bound
          }
        }
    }

                                                                             // --------------
    return max_product;                                                      // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]  upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2]

    return 0;                                             // --------------
}                                                         //  11 + 6*n + [5*n**2, 11*n**2/2  - n/2]  upper bound


Overwriting max_pairwise_product_long_full.cpp


### Count executed intructions in code

In [68]:
%%writefile max_pairwise_product_long_full_execints.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers, long long &execints) {      //INSTRUCTIONS executed by line

    long long max_product = 0LL;                                                          //1
    execints++;

	  long long aux_product = 0LL;                                                          //1
    execints++;

    int n = numbers.size();                                                               //1
    execints++;

    execints++;
    for (int first = 0; first < n; ++first) {                                             //2+2n
        execints += 2;

        execints++;
        for (int second = 0; second < n; ++second) {                                       //2n + 2n**2
          execints += 2;

          execints++;
          if ( first != second ){                                                         //n**2
                aux_product = numbers[first] * numbers[second];                           //n**2 -n
                execints++;

                execints++;
                if (aux_product > max_product){                                           //n**2 -n
                    max_product = aux_product;                                            //[0,n**2/2 - n/2)  upper bound
                    execints++;

                }
        }

      }
      execints++;
    }
    execints++;
                                                                                        // --------------
    return max_product;                                                                 // 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound
}

int main() {                                                 //INSTRUCTIONS executed by line

    long long execints =0LL;

    int n;                                                   //1
    execints++;

    std::cin >> n;                                           //1
    execints++;

    std::vector<int> numbers(n);                             //n
    execints += n;

    execints++;
    for (int i = 0; i < n; ++i) {                            //2 + 2n
        execints += 2;

        std::cin >> numbers[i];
        execints++;                                          //n

    }


    //std::cout << MaxPairwiseProduct(numbers) << "\n";      //1 + 1  + 5 + 2*n + [5*n**2, 11*n**2/2  - n/2] upper bound

    execints++;
    int M = MaxPairwiseProduct(numbers,execints);

    execints++;
    std::cout <<  execints << "\n";
    return 0;                                                // --------------
}                                                            // 11 + 6*n + [5*n**2, 11*n**2/2  - n/2]

Writing max_pairwise_product_long_full_execints.cpp


The number of instructions executed by the algooirithm with a list of size $n$ is  

$$T (n) = 11 + 6n + \left[5n^2, 11\frac{n^2}{2}  - \frac{n}{2}\right)  >  6n^2 = O(n^2)$$

It depends on the size of the lists and  the order of the list.

For the input0.txt
```
10
0 0 0 0 0 0 0 0 0 0
```
$$T = 11 + 6(10) + 5(10)^2   = 561 $$

$$T(10) \in [571, 621)  $$


In [69]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input0.txt

/bin/bash: line 1: input0.txt: No such file or directory


For the input1.txt
```
10
1 2 3 4 5 6 7 8 9 10
```
$$T = 11 + 6(10) + 5(10)^2 +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0  = 601 $$

$$T(10) \in [572, 622)  $$


In [70]:
#The number of times that  max_product = aux_product; in the worst case is less than (n**2 -n)/2 -  upper bound

#     1  2  3  4  5  6  7  8  9  10
#  1  x  2  3  4  5  6  7  8  9  10         9
#  2  2  x  6  8  10 12 14 16 18 20         5 14
#  3  3  6  x  12 15 18 21 24 27 30         4 18
#  4  4  8  12 x  20 24 28 32 36 40         3 21
#  5  5  10 15 20 x  30 35 40 45 50         2 23
#  6  6  12 18 24 30 x  42 48 54 60         2 25
#  7  7  14 21 28 35 42 x  56 63 70         2 27
#  8  8  18 27 36 45 54 63 72 x  90         1 30
#  10 10 20 30 40 50 60 70 60 90 x          0

In [71]:
11 + 6*(10) + 5*(10)*(10) +  9 + 5 + 4 + 3 + 2 + 2 + 2 + 2 + 1 + 0

601

In [72]:
n = 10
print( "[", 11 + 6*n + 5*n**2,  ",", 11 + 6*n  + 5.5*n**2, ")")

[ 571 , 621.0 )


In [73]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input1.txt

600



For the input2.txt
```
10
10 9 8 7 6 5 4 3 2 1
```
$$T = 11 + 6*(10) + 5*(10)**2 +  1  = 571 $$

In [74]:
11 + 6*(10) + 5*(10)**2 +  1

572

In [75]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input2.txt

571


Calculate estimated time by instruction and number of instructions by second

For the input11.txt
```
200000
1 2 ..  199999 2000
```
$$T = 11 + 6*(200000) + 5.5*(200000)**2  = 200001200012  $$

In [76]:

11 + 6*(200000) + 5.5*(200000)**2+1

220001200012.0

In [77]:
!g++ max_pairwise_product_long_full_execints.cpp -o max_pairwise_product_long_full_execints
!./max_pairwise_product_long_full_execints < input11.txt

1
1
2
3
^C


In [78]:
# Estimated time by instruction
# use the real reported  !time ./max_pairwise_product_long_full< input11.txt in seconds intead of the 295

295/200003177092


1.4749765693187072e-09

In [79]:
# Estimated number of instructions by second

200003177092/295

677976871.4983051

### Estimate and verify the maximun n that can be solved in 1 second

$$(5.5n^2 + 5.5n + 11) \times 1.47*10^{-9} s = 1 s$$

$$ n \approx 11000$$


In [80]:
# Verify input13.txt
%%writefile WriteTest.cpp

#include <iostream>
#include <fstream>
#include <cstdlib>

using namespace std;

int main() {
  // Create and open a text file
  ofstream MyFile("input13.txt");

 MyFile << 11000 <<"\n";  // n
 for (int i=1; i <11001; i++){
     MyFile << i <<" ";
 }

  // Close the file
  MyFile.close();
}

Overwriting WriteTest.cpp


In [81]:
!g++ WriteTest.cpp -o WriteTest
!./WriteTest

In [82]:
10999 * 11000

120989000

In [83]:
!g++ max_pairwise_product_long_full.cpp -o max_pairwise_product_long_full
!time ./max_pairwise_product_long_full< input13.txt

120989000

real	0m0.847s
user	0m0.840s
sys	0m0.004s


## Python

In [84]:

%%writefile max_pairwise_product_long_full.py
def MaxPairwiseProduct(numbers):
  max_product = 0
  aux_product = 0
  n = len(numbers)

  for first in range (n):
    for second in range(n):
      if(first != second):
        aux_product = numbers[first]*numbers[second]
        if aux_product > max_product:
          max_product = aux_product
  return max_product

n = int(input())
numbers = [int(i) for i in input().split()]
print(MaxPairwiseProduct(numbers))


Writing max_pairwise_product_long_full.py


In [85]:
# Verify
!time python3 max_pairwise_product_long_full.py < input13.txt

120989000

real	0m19.982s
user	0m19.811s
sys	0m0.024s


# For the follwing solutions rquired  the same tests and analysis performed for the first solution

## Solutions in C++ and Python that uses
$$\frac{n^2 - n}{2} = O(n^2)\ \ \text{products}$$

In [86]:
%%writefile max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,
                numbers[first] * numbers[second]);
        }
    }

    return max_product;
}

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}

Writing max_pairwise_product.cpp


In [87]:
%%writefile max_pairwise_product_long.cpp
#include <iostream>
#include <vector>
#include <algorithm>

long long MaxPairwiseProduct(const std::vector<int>& numbers) {              //INSTRUCTIONS executed by line
    long long max_product = 0LL;                                             //1
	  long long aux_product = 0LL;                                             //1
    int n = numbers.size();                                                  //1

    for (int first = 0; first < n; ++first) {                                //2+2n
        for (int second = first + 1; second < n; ++second) {                 //2n + (n-1)n
			       aux_product =  (long long) numbers[first] * numbers[second];    //(n-1)n/2
            if (max_product < aux_product)  {                                //(n-1)n/2
				        max_product = aux_product;                                   //[1,(n-1)n/2)  upper bound
             }
        }
    }
                                                                             // --------------
    return max_product;                                                      // 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
}

int main() {                                              //INSTRUCTIONS executed by line
    int n;                                                //1
    std::cin >> n;                                        //1
    std::vector<int> numbers(n);                          //n
    for (int i = 0; i < n; ++i) {                         //2 + 2n
        std::cin >> numbers[i];                           //n
    }

    std::cout << MaxPairwiseProduct(numbers) << "\n";     //1 + 1  + 5 + 4n + [2(n-1)n+1, 2.5 (n-1)n) upper bound
    return 0;                                             // --------------
}                                                         // 11 + 7n + [2(n-1)n+1, 2.5 (n-1)n) upper bound


Writing max_pairwise_product_long.cpp


In [88]:
!g++ max_pairwise_product_long.cpp -o max_pairwise_product_long
!time ./max_pairwise_product_long < input13.txt

120989000

real	0m0.594s
user	0m0.576s
sys	0m0.002s


In [89]:
!g++ max_pairwise_product.cpp -o max_pairwise_product
!time ./max_pairwise_product < input13.txt

120989000

real	0m0.568s
user	0m0.562s
sys	0m0.003s


# Sort solution in C++ and Python

$$ O(n \lg n)$$

**Which sorting algorithm is used in GCC?**
https://stackoverflow.com/questions/7221257/which-sorting-algorithm-is-used-in-gcc

* ...  the algorithm is called introsort, a combination between Insertionsort, Quicksort & Heapsort. ...

**Introsort**
https://en.wikipedia.org/wiki/Introsort

Introsort or introspective sort is a hybrid sorting algorithm that provides both fast average performance and (asymptotically) optimal worst-case performance. It begins with quicksort, it switches to heapsort when the recursion depth exceeds a level based on (the logarithm of) the number of elements being sorted and it switches to insertion sort when the number of elements is below some threshold

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Average performance	$O(n \lg n)$

**Tinsort**
https://en.wikipedia.org/wiki/Timsort

Timsort is a hybrid, stable sorting algorithm, derived from merge sort and insertion sort, designed to perform well on many kinds of real-world data. It was implemented by Tim Peters in 2002 for use in the Python programming language.

*   Class	Sorting algorithm
*   Data structure	Array
*   Worst-case performance	$O(n \lg n)$
*   Best-case performance $O(n)$
*   Average performance	$O(n \lg n)$
*   Worst-case space complexity $O(n)$

In [90]:
%%writefile max_pairwise_product_with_sort.py
#Python3
if __name__ == "__main__":
    n = int(input())
    lista = list(map(int, input().split()))
    lista.sort()
    print(lista[n - 1] * lista[n - 2])

Writing max_pairwise_product_with_sort.py


In [91]:
%%writefile max_pairwise_product_with_sort_C.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int main() {
    int n;
    std::cin >> n;
    std::vector<int> numbers(n);
    for (int i = 0; i < n; ++i) {
        std::cin >> numbers[i];
    }

    std::sort(numbers.begin(), numbers.end());
    std::cout << (long long) numbers[n - 1] * numbers[n - 2] << '\n';

    return 0;
}


Overwriting max_pairwise_product_with_sort_C.cpp


In [92]:
!g++ max_pairwise_product_with_sort_C.cpp -o max_pairwise_product_with_sort_C

Test with inputs 1 and 2



In [93]:
!./max_pairwise_product_with_sort_C < input1.txt
!./max_pairwise_product_with_sort_C < input2.txt

90
90


In [94]:
!python max_pairwise_product_with_sort.py < input1.txt
!python max_pairwise_product_with_sort.py < input2.txt

90
90


Test and time it with input 3

In [95]:
!time ./max_pairwise_product_with_sort_C < input3.txt

40000000000

real	0m0.003s
user	0m0.001s
sys	0m0.002s


In [96]:
%%time
!python max_pairwise_product_with_sort.py < input3.txt

40000000000
CPU times: user 6.39 ms, sys: 927 µs, total: 7.32 ms
Wall time: 108 ms


20 stress tests

In [97]:
%%writefile stress_tests.py
#Python3
import random

if __name__ == "__main__":
    for i in range(20):
        n = random.randint(2, 19)
        lista = [random.randint(17, 25631) for i in range(n)]
        lista.sort()
        print(f"n: {n} Product: {lista[n - 1] * lista[n - 2]}")


Writing stress_tests.py


In [159]:
%%writefile stress_tests_C.cpp
#include <iostream>     // Para std::cout
#include <vector>       // Para std::vector
#include <algorithm>    // Para std::sort
#include <random>       // Para std::random_device, std::mt19937, std::uniform_int_distribution

int main() {
    std::random_device rd;                      // Generador de semilla
    std::mt19937 gen(rd());                     // Generador de números aleatorios
    std::uniform_int_distribution<int> n_dist(2, 99);      // Distribución para n
    std::uniform_int_distribution<int> num_dist(0, 25000); // Distribución para los números

    for (int i = 0; i < 100; ++i) {
        int n = n_dist(gen);                    // Generar tamaño aleatorio del vector
        std::vector<int> numbers(n);           // Vector de tamaño n
        for (int j = 0; j < n; ++j) {          // Llenar el vector con números aleatorios
            numbers[j] = num_dist(gen);
        }

        std::sort(numbers.begin(), numbers.end()); // Ordenar el vector
        std::cout << "n= " << n
                  << " Product= " << (long long) numbers[n - 1] * numbers[n - 2] << '\n';
    }

    return 0;
}


Overwriting stress_tests_C.cpp


In [160]:
!g++ stress_tests_C.cpp -o stress_tests_C
!./stress_tests_C


n= 23 Product= 595846209
n= 22 Product= 570633000
n= 51 Product= 608230164
n= 46 Product= 610261352
n= 49 Product= 618629046
n= 17 Product= 468046062
n= 65 Product= 620281380
n= 68 Product= 555757328
n= 56 Product= 594852700
n= 40 Product= 565320146
n= 21 Product= 585676692
n= 86 Product= 599098896
n= 66 Product= 604643400
n= 86 Product= 608838146
n= 32 Product= 570301720
n= 33 Product= 588470520
n= 29 Product= 540879096
n= 40 Product= 618950000
n= 82 Product= 619801015
n= 88 Product= 620826860
n= 27 Product= 455116890
n= 91 Product= 601937196
n= 43 Product= 536770080
n= 14 Product= 531455120
n= 67 Product= 592159100
n= 11 Product= 556265178
n= 2 Product= 12248514
n= 68 Product= 547745772
n= 84 Product= 603802053
n= 98 Product= 617343732
n= 75 Product= 614163965
n= 69 Product= 618130100
n= 58 Product= 566744988
n= 19 Product= 566733839
n= 47 Product= 594097264
n= 36 Product= 564833786
n= 51 Product= 604681134
n= 59 Product= 588777904
n= 65 Product= 604457235
n= 10 Product= 388962705
n=

In [100]:
!python stress_tests.py

n: 18 Product: 470538420
n: 13 Product: 591527195
n: 7 Product: 300331314
n: 17 Product: 555544611
n: 16 Product: 477813172
n: 5 Product: 508714528
n: 9 Product: 571645890
n: 10 Product: 555204300
n: 13 Product: 553726080
n: 3 Product: 367129742
n: 17 Product: 373796034
n: 11 Product: 426178618
n: 17 Product: 584402910
n: 9 Product: 392295204
n: 13 Product: 610532745
n: 4 Product: 367944537
n: 13 Product: 530496824
n: 5 Product: 392158450
n: 7 Product: 515088359
n: 18 Product: 538579500


Count executed instructions in code and verify with inputs 1,2 and 3


In [101]:
%%writefile max_pairwise_product_with_sort_and_instructions.py
#Python3
import math
if __name__ == "__main__":
    inst=0
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+= 4*n
    lista.sort()
    inst+=  n * math.log2(n)
    print(lista[n - 1] * lista[n - 2])
    inst+=1
    print(f"Instrucciones ejecutadas: {int(inst)}\n")

Writing max_pairwise_product_with_sort_and_instructions.py


In [161]:
%%writefile max_pairwise_product_with_sort_C_and_instructions.cpp
#include <iostream>     // Para std::cin y std::cout
#include <vector>       // Para std::vector
#include <algorithm>    // Para std::sort
#include <cmath>        // Para log2

int main() {
    int inst = 0;  // Contador de instrucciones
    int n;         // Número de elementos
    inst++;        // Incrementar contador por la declaración de inst y n
    std::cin >> n; // Leer el número de elementos
    inst++;
    std::vector<int> numbers(n); // Crear un vector de tamaño n
    inst += n;  // Considerar las instrucciones para inicializar el vector
    inst++;     // Declaración de i=0

    for (int i = 0; i < n; ++i) { // Bucle para leer los números
        inst += 2;                // Comparación y actualización del índice
        std::cin >> numbers[i];   // Leer el elemento i del vector
        inst++;
    }

    std::sort(numbers.begin(), numbers.end()); // Ordenar el vector
    inst += n * std::log2(n);                  // Instrucciones ejecutadas por el ordenamiento

    std::cout << (long long) numbers[n - 1] * numbers[n - 2] << '\n'; // Imprimir el producto máximo
    inst += 2; // Operaciones de salida

    std::cout << "Instrucciones Ejecutadas: " << inst << '\n'; // Imprimir el total de instrucciones
    return 0;
}


Overwriting max_pairwise_product_with_sort_C_and_instructions.cpp


In [162]:
!g++ max_pairwise_product_with_sort_C_and_instructions.cpp -o max_pairwise_product_with_sort_C_and_instructions

In [163]:
!./max_pairwise_product_with_sort_C_and_instructions < input1.txt
!./max_pairwise_product_with_sort_C_and_instructions < input2.txt
!./max_pairwise_product_with_sort_C_and_instructions < input3.txt

90
Instrucciones Ejecutadas: 78
90
Instrucciones Ejecutadas: 78
40000000000
Instrucciones Ejecutadas: 44


In [105]:
!python max_pairwise_product_with_sort_and_instructions.py < input1.txt
!python max_pairwise_product_with_sort_and_instructions.py < input2.txt
!python max_pairwise_product_with_sort_and_instructions.py < input3.txt

90
Instrucciones ejecutadas: 75

90
Instrucciones ejecutadas: 75

40000000000
Instrucciones ejecutadas: 41



Calculate estimated time by instruction and number of instructions by second

In [164]:
%%writefile max_pairwise_product_with_sort_and_instructions_with_time_and_IbS.py
#Python3
import math
import time
if __name__ == "__main__":
    inicio = time.time()
    inst=0
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+= 4*n
    lista.sort()
    inst+=  n * math.log2(n)
    print(lista[n - 1] * lista[n - 2])
    inst+=1
    print(f"Instrucciones ejecutadas: {int(inst)}")
    fin = time.time()
    tiempo_total = fin - inicio
    instructionTime = tiempo_total/inst
    instructionBySecond= inst/tiempo_total

    print(f"Tiempo total de ejecución: {tiempo_total*1000} milisegundos")
    print(f"Tiempo estimado de cada instruccion: {instructionTime*1000} milisegundos")
    print(f"Instrucciones por segundo: {int(instructionBySecond)} ")

Overwriting max_pairwise_product_with_sort_and_instructions_with_time_and_IbS.py


In [107]:
!python max_pairwise_product_with_sort_and_instructions_with_time_and_IbS.py < input13.txt

120989000
Instrucciones ejecutadas: 191679
Tiempo total de ejecución: 2.4878978729248047 milisegundos
Tiempo estimado de cada instruccion: 1.2979476136917105e-05 milisegundos
Instrucciones por segundo: 77044711 


### Estimate and verify the maximun n that can be solved in 1 second

In [108]:
%%writefile maximun_n_that_can_be_solved.py
#Python3
import random

if __name__ == "__main__":
      n = 600000
      lista = [random.randint(17, 25631) for i in range(n)]
      lista.sort()
      print(f"n: {n} Product: {lista[n - 1] * lista[n - 2]}")

Writing maximun_n_that_can_be_solved.py


In [109]:
%%time
!python maximun_n_that_can_be_solved.py

n: 600000 Product: 656948161
CPU times: user 14.5 ms, sys: 1.05 ms, total: 15.6 ms
Wall time: 911 ms


In [166]:
%%writefile maximun_n_that_can_be_solved_C.cpp
#include <iostream>     // Para std::cout
#include <vector>       // Para std::vector
#include <algorithm>    // Para std::sort
#include <random>       // Para std::random_device, std::mt19937, std::uniform_int_distribution

int main() {
    std::random_device rd;                          // Generador de semilla
    std::mt19937 gen(rd());                         // Generador de números aleatorios
    std::uniform_int_distribution<int> num_dist(0, 25000); // Distribución para números

    int n = 1700000;                                // Tamaño del vector
    std::vector<int> numbers(n);                    // Crear un vector de tamaño n
    for (int i = 0; i < n; ++i) {                   // Llenar el vector con números aleatorios
        numbers[i] = num_dist(gen);
    }

    std::sort(numbers.begin(), numbers.end());      // Ordenar el vector

    // Imprimir el producto de los dos números más grandes
    std::cout << "n= " << n
              << " Product= " << (long long) numbers[n - 1] * numbers[n - 2] << '\n';

    return 0;
}


Overwriting maximun_n_that_can_be_solved_C.cpp


In [167]:
!g++ maximun_n_that_can_be_solved_C.cpp -o maximun_n_that_can_be_solved_C
!time ./maximun_n_that_can_be_solved_C

n= 1700000 Product= 625000000

real	0m0.893s
user	0m0.882s
sys	0m0.008s


## Efficient solution with largest and second-largest in an list, array or vector in C++ and Python

 $$T(n) < 10 n  = O(n)$$





Add in each line a coment with the number of intstrcution that are executed in that line

In [168]:
%%writefile max_pairwise_product_long_efficient.cpp
#include <iostream>     // Para std::cin y std::cout
#include <vector>       // Para std::vector
#include <algorithm>    // Opcional, si se necesitara en el futuro

// Función para encontrar el producto máximo de dos elementos en un vector
long long MaxPairwiseProduct(const std::vector<long long>& numbers) {
    long long max_product = 0LL;          // Producto máximo
    long long n_major = 0LL;              // Mayor número encontrado
    long long n_major_2 = 0LL;            // Segundo mayor número encontrado
    int n = numbers.size();               // Tamaño del vector

    for (int i = 0; i < n; ++i) {         // Iterar a través del vector
        if (numbers[i] > n_major) {
            n_major_2 = n_major;          // Actualizar el segundo mayor
            n_major = numbers[i];         // Actualizar el mayor
        } else if (numbers[i] > n_major_2) {
            n_major_2 = numbers[i];       // Actualizar el segundo mayor si corresponde
        }
    }
    max_product = n_major * n_major_2;    // Calcular el producto máximo
    return max_product;
}

int main() {
    int n;                                // Número de elementos en el vector
    std::cin >> n;                        // Leer el tamaño del vector
    std::vector<long long> numbers(n);    // Crear un vector de tamaño n

    for (int i = 0; i < n; ++i) {         // Leer los elementos del vector
        std::cin >> numbers[i];
    }

    // Calcular y mostrar el producto máximo
    std::cout << MaxPairwiseProduct(numbers) << "\n";
    return 0;
}


Overwriting max_pairwise_product_long_efficient.cpp


In [169]:
!g++ max_pairwise_product_long_efficient.cpp -o max_pairwise_product_long_efficient

In [114]:
%%writefile max_pairwise_product_efficient.py
#Python3
def max_pairwise_product(lista):
    n_major = 0
    n_major_2 = 0
    max_product = 0
    for i in lista:
        if i > n_major:
            n_major_2=n_major
            n_major = i
        elif i > n_major_2:
            n_major_2=i
    max_product = n_major_2 * n_major

    return max_product

if __name__ == "__main__":
    n = int(input())
    lista = list(map(int, input().split()))
    print(max_pairwise_product(lista))

Writing max_pairwise_product_efficient.py


Test with inputs 1 and 2



In [170]:
!./max_pairwise_product_long_efficient < input1.txt
!./max_pairwise_product_long_efficient < input2.txt

90
90


In [171]:
!python max_pairwise_product_efficient.py < input1.txt
!python max_pairwise_product_efficient.py < input2.txt

90
90


Test and time it with input 3

In [117]:
%%time
!python max_pairwise_product_efficient.py < input3.txt

40000000000
CPU times: user 5.73 ms, sys: 2.76 ms, total: 8.48 ms
Wall time: 109 ms


In [118]:
!time ./max_pairwise_product_long_efficient < input3.txt

/bin/bash: line 1: ./max_pairwise_product_long_efficient: No such file or directory

real	0m0.001s
user	0m0.001s
sys	0m0.000s


20 stress tests

In [186]:
%%writefile max_pairwise_product_long_efficient_stress_test.cpp
#include <iostream>     // Para std::cin, std::cout
#include <vector>       // Para std::vector
#include <random>       // Para la generación de números aleatorios

long long MaxPairwiseProduct(const std::vector<int>& numbers) {
    long long max_product = 0LL;  // 1
    long long n_major = 0LL;      // 1
    long long n_major_2 = 0LL;    // 1
    int n = numbers.size();       // 1

    for (int i = 0; i < n; ++i) {  // 1 + 2n
        if (numbers[i] > n_major) {
            n_major_2 = n_major;  // n
            n_major = numbers[i];
        }
        else if (numbers[i] > n_major_2) {  // n
            n_major_2 = numbers[i];  // [0, n]
        }
    }
    max_product = n_major * n_major_2;  // --------------
    return max_product;
}

int main() {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution<int> n_dist(2, 19);     // Rango para n
    std::uniform_int_distribution<int> num_dist(0, 25000); // Rango para los números

    for (int i = 0; i < 20; ++i) {
        int n = n_dist(gen);  // Genera un número aleatorio para n
        std::vector<int> numbers(n);  // Crea el vector con tamaño n

        for (int i = 0; i < n; ++i) {  // 1 + 2n
            numbers[i] = num_dist(gen);  // Llenar el vector con números aleatorios
        }

        std::cout << "n= " << n << " Product= " << MaxPairwiseProduct(numbers) << "\n";
    }

    return 0;
}


Overwriting max_pairwise_product_long_efficient_stress_test.cpp


In [187]:
!g++ max_pairwise_product_long_efficient_stress_test.cpp -o max_pairwise_product_long_efficient_stress_test
!./max_pairwise_product_long_efficient_stress_test

n= 18 Product= 534887534
n= 19 Product= 569558600
n= 16 Product= 578000864
n= 10 Product= 582480150
n= 4 Product= 134033140
n= 13 Product= 600635385
n= 2 Product= 15343002
n= 13 Product= 491234778
n= 19 Product= 565243229
n= 15 Product= 372708405
n= 9 Product= 464656560
n= 12 Product= 503468891
n= 8 Product= 510359856
n= 14 Product= 392748207
n= 2 Product= 320874345
n= 3 Product= 333137408
n= 17 Product= 395998554
n= 14 Product= 512064878
n= 2 Product= 124859108
n= 13 Product= 355831920


In [121]:
%%writefile max_pairwise_product_efficient_stress_tests.py
#Python3
import random
def max_pairwise_product(lista):
    n_major = 0
    n_major_2 = 0
    max_product = 0
    for i in lista:
        if i > n_major:
            n_major_2=n_major
            n_major = i
        elif i > n_major_2:
            n_major_2=i
    max_product = n_major_2 * n_major

    return max_product

if __name__ == "__main__":
    for i in range(20):
      n = random.randint(2, 19)
      lista = [random.randint(17, 25631) for i in range(n)]
      print(f"n=  {n}   Product= {max_pairwise_product(lista)}")

Writing max_pairwise_product_efficient_stress_tests.py


In [122]:
!python max_pairwise_product_efficient_stress_tests.py

n=  13   Product= 566704342
n=  14   Product= 552106628
n=  9   Product= 465240360
n=  17   Product= 566125240
n=  16   Product= 543382985
n=  16   Product= 582945390
n=  12   Product= 501524310
n=  18   Product= 597477198
n=  14   Product= 617183490
n=  3   Product= 329533876
n=  17   Product= 581775432
n=  4   Product= 313664612
n=  16   Product= 628655248
n=  19   Product= 543979816
n=  17   Product= 464161472
n=  13   Product= 570679884
n=  16   Product= 580030407
n=  17   Product= 517303449
n=  13   Product= 530937434
n=  5   Product= 298310298


Count executed instructions in code and verify with inputs 1,2 and 3


In [183]:
%%writefile max_pairwise_product_long_efficient_and_instructions.cpp
#include <iostream>     // Para std::cin y std::cout
#include <vector>       // Para std::vector

long long inst = 0LL;

long long MaxPairwiseProduct(const std::vector<long long>& numbers) {
    long long max_product = 0LL;  // 1
    inst++;
    long long n_major = 0LL;      // 1
    inst++;
    long long n_major_2 = 0LL;   // 1
    inst++;
    int n = numbers.size();       // 1
    inst++;

    for(int i = 0; i < n; ++i) {  // 1 + 2n
        inst += 2;
        if (numbers[i] > n_major) {
            inst++;
            n_major_2 = n_major;  // n
            inst++;
            n_major = numbers[i];
            inst++;
        } else if (numbers[i] > n_major_2) {
            inst++;
            n_major_2 = numbers[i];  // n
            inst++;
        }
    }
    max_product = n_major * n_major_2;
    inst++;  // --------------
    return max_product;
}

int main() {
    int n;  // 1
    inst++;
    std::cin >> n;  // 1
    inst++;
    std::vector<long long> numbers(n);  // n
    inst++;

    for (int i = 0; i < n; ++i) {  // 1 + 2n
        std::cin >> numbers[i];  // n
    }
    inst += n;

    std::cout << MaxPairwiseProduct(numbers) << "\n";  // Mostrar el producto máximo
    inst++;
    std::cout << "Numero de instrucciones: " << inst << "\n";  // Número de instrucciones ejecutadas
    return 0;
}


Overwriting max_pairwise_product_long_efficient_and_instructions.cpp


In [184]:
!g++ max_pairwise_product_long_efficient_and_instructions.cpp -o max_pairwise_product_long_efficient_and_instructions

In [185]:
!./max_pairwise_product_long_efficient_and_instructions < input1.txt
!./max_pairwise_product_long_efficient_and_instructions < input2.txt
!./max_pairwise_product_long_efficient_and_instructions < input3.txt

90
Numero de instrucciones: 69
90
Numero de instrucciones: 44
40000000000
Numero de instrucciones: 40


In [126]:
%%writefile max_pairwise_product_efficient_and_instructions.py
#Python3

inst=0
def max_pairwise_product(lista):
    global inst
    n_major = 0
    inst+=1
    n_major_2 = 0
    inst+=1
    max_product = 0
    inst+=1
    for i in lista:
        inst+=1
        if i > n_major:
            inst+=1
            n_major_2=n_major
            inst+=1
            n_major = i
            inst+=1
        elif i > n_major_2:
            inst+=1
            n_major_2=i
            inst+=1
            inst+=1
    max_product = n_major_2 * n_major
    inst+=1

    return max_product

if __name__ == "__main__":
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+=4*n
    print(max_pairwise_product(lista))
    inst+=1
    print(f"Numero Total de Instrucciones:  {inst}")


Writing max_pairwise_product_efficient_and_instructions.py


In [127]:
!python max_pairwise_product_efficient_and_instructions.py < input1.txt
!python max_pairwise_product_efficient_and_instructions.py < input2.txt
!python max_pairwise_product_efficient_and_instructions.py < input3.txt

90
Numero Total de Instrucciones:  86
90
Numero Total de Instrucciones:  62
40000000000
Numero Total de Instrucciones:  51


Calculate estimated time by instruction and number of instructions by second

In [180]:
%%writefile max_pairwise_product_long_efficient_and_instructions_by_second.cpp
#include <iostream>     // Para std::cin y std::cout
#include <vector>       // Para std::vector
#include <chrono>       // Para medir el tiempo de ejecución

long long inst = 0LL;

long long MaxPairwiseProduct(const std::vector<long long>& numbers) {
    long long max_product = 0LL;  // 1
    inst++;
    long long n_major = 0LL;      // 1
    inst++;
    long long n_major_2 = 0LL;   // 1
    inst++;
    int n = numbers.size();       // 1
    inst++;

    for(int i = 0; i < n; ++i) {  // 1 + 2n
        inst += 2;
        if (numbers[i] > n_major) {
            inst++;
            n_major_2 = n_major;  // n
            inst++;
            n_major = numbers[i];
            inst++;
        } else if (numbers[i] > n_major_2) {
            inst++;
            n_major_2 = numbers[i];  // n
            inst++;
        }
    }
    max_product = n_major * n_major_2;
    inst++;  // --------------
    return max_product;
}

int main() {
    auto start_time = std::chrono::high_resolution_clock::now();  // Medir el tiempo de inicio

    int n;  // 1
    inst++;
    std::cin >> n;  // 1
    inst++;
    std::vector<long long> numbers(n);  // n
    inst++;

    for (int i = 0; i < n; ++i) {  // 1 + 2n
        std::cin >> numbers[i];  // n
    }
    inst += n;

    std::cout << MaxPairwiseProduct(numbers) << "\n";  // Mostrar el producto máximo
    inst++;

    auto end_time = std::chrono::high_resolution_clock::now();  // Medir el tiempo de fin
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time);  // Duración en milisegundos

    float duration_sec, duration_f;
    duration_f = static_cast<float>(duration.count());
    duration_sec = duration_f / 1000;

    std::cout << "Numero de instrucciones: " << inst << "\n";
    std::cout << "Duracion Estimada: " << duration_sec << " segundos\n";
    std::cout << "Duracion Estimada por instrucción: " << duration_sec / inst << " segundos\n";
    std::cout << "Numero de instrucciones por segundo: " << 1 / (duration_sec / inst) << " Instrucciones\n";

    return 0;
}


Overwriting max_pairwise_product_long_efficient_and_instructions_by_second.cpp


In [181]:
!g++ max_pairwise_product_long_efficient_and_instructions_by_second.cpp -o max_pairwise_product_long_efficient_and_instructions_by_second


In [182]:
!./max_pairwise_product_long_efficient_and_instructions_by_second < input13.txt

120989000
Numero de instrucciones: 66009
Duracion Estimada: 0.006 segundos
Duracion Estimada por instrucción: 9.08967e-08 segundos
Numero de instrucciones por segundo: 1.10015e+07 Instrucciones


In [131]:
%%writefile max_pairwise_product_efficient_and_instructions_by_second.py
#Python3
import time

inst=0
def max_pairwise_product(lista):
    global inst
    n_major = 0
    inst+=1
    n_major_2 = 0
    inst+=1
    max_product = 0
    inst+=1
    for i in lista:
        inst+=1
        if i > n_major:
            inst+=1
            n_major_2=n_major
            inst+=1
            n_major = i
            inst+=1
        elif i > n_major_2:
            inst+=1
            n_major_2=i
            inst+=1
            inst+=1
    max_product = n_major_2 * n_major
    inst+=1

    return max_product

if __name__ == "__main__":
    inicio=time.time()
    n = int(input())
    inst+=1
    lista = list(map(int, input().split()))
    inst+=4*n
    print(max_pairwise_product(lista))
    inst+=1
    fin = time.time()
    tiempo_total = fin - inicio
    instructionTime = tiempo_total/inst
    instructionBySecond= 1/instructionTime
    print(f"Numero Total de Instrucciones:  {inst}")
    print(f"Tiempo total de ejecución: {tiempo_total} segundos")
    print(f"Tiempo estimado de cada instruccion: {instructionTime} segundos")
    print(f"Instrucciones por segundo: {int(instructionBySecond)} ")


Writing max_pairwise_product_efficient_and_instructions_by_second.py


In [132]:
!python max_pairwise_product_efficient_and_instructions_by_second.py < input13.txt

120989000
Numero Total de Instrucciones:  88006
Tiempo total de ejecución: 0.01212620735168457 segundos
Tiempo estimado de cada instruccion: 1.377884161498599e-07 segundos
Instrucciones por segundo: 7257504 


### Estimate and verify the maximun n that can be solved in 1 second

In [133]:
%%writefile maximun_n_that_can_be_solved_efficient.py
#Python3
import random
def max_pairwise_product(lista):
    n_major = 0
    n_major_2 = 0
    max_product = 0
    for i in lista:
        if i > n_major:
            n_major_2=n_major
            n_major = i
        elif i > n_major_2:
            n_major_2=i
    max_product = n_major_2 * n_major

    return max_product

if __name__ == "__main__":
    n = 800000
    lista = [random.randint(17, 25631) for i in range(n)]
    print(f"n=  {n}   Product= {max_pairwise_product(lista)}")

Writing maximun_n_that_can_be_solved_efficient.py


In [134]:
%%time
!python maximun_n_that_can_be_solved_efficient.py

n=  800000   Product= 656948161
CPU times: user 14.3 ms, sys: 1.28 ms, total: 15.6 ms
Wall time: 1.01 s


## Super efficient solution

Whitout storing the list in memory, only stpring the largest and second largest and whitout calling a function, add in each line a coment with the number of instructions that are executed in that line

in C++ and Python


 $$T(n) < 13 n  = O(n)$$


In [177]:
%%writefile max_pairwise_super_efficient.cpp
#include <iostream>  // Para std::cin y std::cout

using namespace std;

int main() {
    int n;                                     // 1
    cin >> n;                                  // 1
    int numero_M = 0;                          // 1
    int segundo_Numero_M = 0;                  // 1

    // Iterar sobre todos los elementos para encontrar los dos mayores números
    for (int i = 0; i < n; ++i) {              // 1 + 2n
        int comp = 0;                          // n
        cin >> comp;                           // n

        if (comp > numero_M) {                 // n
            segundo_Numero_M = numero_M;       // n
            numero_M = comp;                   // n
        } else if (comp > segundo_Numero_M) {  // n
            segundo_Numero_M = comp;           // n
        }
    }

    cout << numero_M * segundo_Numero_M << "\n"; // 1
    return 0;
}


Overwriting max_pairwise_super_efficient.cpp


In [178]:
!g++ max_pairwise_super_efficient.cpp -o max_pairwise_super_efficient

Test with inputs 1 and 2




In [179]:
!./max_pairwise_super_efficient < input1.txt
!./max_pairwise_super_efficient < input2.txt

90
90


Test and time it with input 3

In [138]:
!time ./max_pairwise_super_efficient < input3.txt

/bin/bash: line 1: ./max_pairwise_super_efficient: No such file or directory

real	0m0.001s
user	0m0.000s
sys	0m0.001s


20 stress tests

In [175]:
%%writefile max_pairwise_super_efficient_stress_tests.cpp
#include <iostream>     // Para std::cout y std::cin
#include <vector>       // Para std::vector
#include <random>       // Para std::random_device, std::mt19937, std::uniform_int_distribution

using namespace std;

int main() {
    std::random_device rd;                          // Generador de semilla
    std::mt19937 gen(rd());                         // Generador de números aleatorios
    std::uniform_int_distribution<int> n_dist(2, 19); // Distribución para el tamaño del arreglo
    std::uniform_int_distribution<int> num_dist(0, 25000); // Distribución para los números

    // Realizar 20 pruebas de estrés
    for (int i = 0; i < 20; ++i) {
        int n = n_dist(gen);                        // Tamaño del arreglo
        int numero_M = 0;                           // Mayor número
        int segundo_Numero_M = 0;                   // Segundo mayor número

        // Llenar el arreglo con números aleatorios y calcular el producto
        for (int i = 0; i < n; ++i) {
            int comp = num_dist(gen);              // Generar un número aleatorio
            if (comp > numero_M) {                 // Si el número es mayor que el mayor conocido
                segundo_Numero_M = numero_M;      // Actualizar el segundo mayor
                numero_M = comp;                  // Actualizar el mayor
            } else if (comp > segundo_Numero_M) {  // Si el número es mayor que el segundo mayor
                segundo_Numero_M = comp;          // Actualizar el segundo mayor
            }
        }

        // Mostrar el tamaño y el producto de los dos mayores números
        cout << "n= " << n << " Multiplicacion= " << numero_M * segundo_Numero_M  << "\n";
    }

    return 0;
}


Overwriting max_pairwise_super_efficient_stress_tests.cpp


In [176]:
!g++ max_pairwise_super_efficient_stress_tests.cpp -o max_pairwise_super_efficient_stress_tests
!./max_pairwise_super_efficient_stress_tests

n= 13 Multiplicacion= 396979086
n= 4 Multiplicacion= 281272461
n= 8 Multiplicacion= 396370680
n= 9 Multiplicacion= 527312328
n= 16 Multiplicacion= 593965851
n= 7 Multiplicacion= 354469400
n= 16 Multiplicacion= 501989101
n= 14 Multiplicacion= 562635942
n= 10 Multiplicacion= 553657651
n= 3 Multiplicacion= 556180371
n= 14 Multiplicacion= 548939308
n= 4 Multiplicacion= 354149865
n= 5 Multiplicacion= 384987429
n= 6 Multiplicacion= 178297728
n= 17 Multiplicacion= 443930397
n= 16 Multiplicacion= 515733284
n= 19 Multiplicacion= 576488108
n= 18 Multiplicacion= 478287022
n= 14 Multiplicacion= 583085040
n= 4 Multiplicacion= 556350976


Count executed instructions in code and verify with inputs 1,2 and 3

In [172]:
%%writefile max_pairwise_super_efficient_count.cpp
#include <iostream> // Para std::cin y std::cout
using namespace std;

int main() {
    int inst = 0;                        // Contador de instrucciones
    int n;                               // Número de elementos
    inst++;
    cin >> n;                            // Leer el tamaño del arreglo
    inst++;

    int numero_M = 0;                    // Mayor número
    inst++;
    int segundo_Numero_M = 0;            // Segundo mayor número
    inst++;

    inst++;                              // Inicialización del bucle
    inst += 2 * n;                       // Por las iteraciones del bucle
    for (int i = 0; i < n; ++i) {        // Iterar sobre el arreglo
        int comp = 0;                    // Número actual
        inst++;
        cin >> comp;                     // Leer el número actual
        inst++;
        inst++;
        if (comp > numero_M) {           // Si el número actual es mayor que el mayor conocido
            segundo_Numero_M = numero_M; // Actualizar el segundo mayor
            inst++;
            numero_M = comp;             // Actualizar el mayor
            inst++;
        } else if (comp > segundo_Numero_M) { // Si el número actual es mayor que el segundo mayor
            inst++;
            segundo_Numero_M = comp;     // Actualizar el segundo mayor
            inst++;
        }
    }

    cout << numero_M * segundo_Numero_M << "\n"; // Producto del mayor y segundo mayor
    inst++;
    cout << "Numero de Instrucciones: " << inst << "\n"; // Imprimir el conteo de instrucciones
    return 0;
}


Overwriting max_pairwise_super_efficient_count.cpp


In [173]:
!g++ max_pairwise_super_efficient_count.cpp -o max_pairwise_super_efficient_count

In [174]:
!./max_pairwise_super_efficient_count < input1.txt
!./max_pairwise_super_efficient_count < input2.txt
!./max_pairwise_super_efficient_count < input3.txt

90
Numero de Instrucciones: 76
90
Numero de Instrucciones: 60
1345294336
Numero de Instrucciones: 46


Calculate estimated time by instruction and number of instructions by second

In [144]:
t_por_instruccion = 0.003 / 46
instrucciones_por_segundo = 1 / t_por_instruccion
print(f"Tiempo por instruccion: {t_por_instruccion}")
print(f"Instrucciones por segundo: {instrucciones_por_segundo}")

Tiempo por instruccion: 6.521739130434783e-05
Instrucciones por segundo: 15333.333333333332


### Super efficient solution in Python

Whitout storing the list in memory and whitout calling a function, add in each line a coment with the number of intstrcution that are executed in that line

In [145]:

%%writefile max_pairwise_super_efficient.py
n = int(input())                            # 1
numero_M = 0                                # 1
segundo_Numero_M = 0                        # 1

for comp in input().split():                 # 3 + 2n

    comp = int(comp)                          # n

    if comp > numero_M:                    # n
        segundo_Numero_M = numero_M      # 0 o n
        numero_M = comp                  # 0 o n
    elif comp > segundo_Numero_M:           # n
        segundo_Numero_M = comp             # 0 o n
print(f"{numero_M * segundo_Numero_M}")

Writing max_pairwise_super_efficient.py


Test with inputs 1 and 2




In [146]:
!python max_pairwise_super_efficient.py < input1.txt
!python max_pairwise_super_efficient.py < input2.txt

90
90


Test and time it with input 3

In [147]:
%%time
!python max_pairwise_super_efficient.py < input3.txt

40000000000
CPU times: user 7.86 ms, sys: 998 µs, total: 8.86 ms
Wall time: 105 ms


20 stress tests

In [148]:
%%writefile max_pairwise_super_efficient_stress_tests.py
import random

for i in range(20):
  n = random.randint(2, 19)
  numero_M = 0                                # 1
  segundo_Numero_M = 0
  for j in range(n):
    comp=random.randint(17, 25631)
    if comp > numero_M:                    # n
      segundo_Numero_M = numero_M      # 0 o n
      numero_M = comp                  # 0 o n
    elif comp > segundo_Numero_M:           # n
      segundo_Numero_M = comp             # 0 o n
  print(f"{i+1}. n= {n} Producto= {numero_M * segundo_Numero_M}")

Writing max_pairwise_super_efficient_stress_tests.py


In [149]:
!python max_pairwise_super_efficient_stress_tests.py

1. n= 8 Producto= 517111596
2. n= 10 Producto= 587190093
3. n= 14 Producto= 626635213
4. n= 6 Producto= 462784393
5. n= 2 Producto= 62283312
6. n= 16 Producto= 477438990
7. n= 17 Producto= 530141850
8. n= 5 Producto= 556708578
9. n= 14 Producto= 615321731
10. n= 11 Producto= 511319672
11. n= 2 Producto= 142076800
12. n= 2 Producto= 209123200
13. n= 6 Producto= 304888155
14. n= 4 Producto= 367242564
15. n= 14 Producto= 469674063
16. n= 11 Producto= 614074875
17. n= 6 Producto= 230108260
18. n= 19 Producto= 550194052
19. n= 2 Producto= 215708612
20. n= 6 Producto= 481224060


Count executed instructions in code and verify with inputs 1,2 and 3

In [150]:
%%writefile max_pairwise_super_efficient_count.py
instrucciones=0
n = int(input())                            # 1
instrucciones+=1
numero_M = 0                                # 1
instrucciones+=1
segundo_Numero_M = 0                        # 1
instrucciones+=1
instrucciones+=1
instrucciones+=2*n
for comp in input().split():                 # 1 + 2n

    comp = int(comp)                          # n
    instrucciones+=1

    instrucciones+=1
    if comp > numero_M:                    # n
        segundo_Numero_M = numero_M      # 0 o n
        instrucciones+=1
        numero_M = comp                  # 0 o n
        instrucciones+=1
    elif comp > segundo_Numero_M:           # 0 o n
        instrucciones+=1
        segundo_Numero_M = comp             # 0 o n
        instrucciones+=1

print(f"{numero_M * segundo_Numero_M}")
print(f"Instrucciones totales: {instrucciones} ")

Writing max_pairwise_super_efficient_count.py


In [151]:
!python max_pairwise_super_efficient_count.py < input1.txt
!python max_pairwise_super_efficient_count.py < input2.txt
!python max_pairwise_super_efficient_count.py < input3.txt

90
Instrucciones totales: 64 
90
Instrucciones totales: 48 
40000000000
Instrucciones totales: 38 


## Calculate estimated time by instruction and number of instructions by second

In [152]:
t_por_instruccion = 0.00405 / 38
instrucciones_por_segundo = 1 / t_por_instruccion
print(f"Tiempo por instruccion: {t_por_instruccion}")
print(f"Instrucciones por segundo: {instrucciones_por_segundo}")

Tiempo por instruccion: 0.00010657894736842105
Instrucciones por segundo: 9382.716049382716


### Estimate and verify the maximun n that can be solved in 1 second

In [153]:

%%writefile maximun_n_that_can_be_solved_super_efficient.py
import random

n = 900000
numero_M = 0                                # 1
segundo_Numero_M = 0
for j in range(n):
  comp=random.randint(17, 25631)
  if comp > numero_M:                    # n
    segundo_Numero_M = numero_M      # 0 o n
    numero_M = comp                  # 0 o n
  elif comp > segundo_Numero_M:           # n
    segundo_Numero_M = comp             # 0 o n
print(f"Producto= {numero_M * segundo_Numero_M}")

Writing maximun_n_that_can_be_solved_super_efficient.py


In [154]:
%%time
!python maximun_n_that_can_be_solved_super_efficient.py

Producto= 656948161
CPU times: user 22.9 ms, sys: 1.46 ms, total: 24.4 ms
Wall time: 1.81 s


## Perform 100 randomized automated tests in C++ and Python with a porgram  compares the answers,for the same 100 random input instances, of the super efficient solution $(T(n) < 10 n = O(n))$ against the sorting solution  $(T(n)  = O(n \lg{n}))$,  if the answers are not the same stops and displays "Error" listing  the input and the two answers; if the 100 tests are completed without stoping displays "Pass".

In [155]:
%%writefile max_pairwise_super_efficient_against_sorting_stress_tests.py
import random

for i in range(100):
  n = random.randint(2, 99)
  numero_M = 0                                # 1
  segundo_Numero_M = 0
  lista = []
  for j in range(n):
    comp=random.randint(17, 25631)
    lista.append(comp)
    if comp > numero_M:                    # n
      segundo_Numero_M = numero_M      # 0 o n
      numero_M = comp                  # 0 o n
    elif comp > segundo_Numero_M:           # n
      segundo_Numero_M = comp             # 0 o n
  lista.sort()
  sol_sorting=lista[n - 1] * lista[n - 2]
  sol_super_e = numero_M * segundo_Numero_M


  if sol_sorting != sol_super_e:
    print("Error")
    break
  else:
      print(f"{i+1}. SUPER EFFICENT SOLITION: n= {n} Product= {numero_M * segundo_Numero_M}")
      print(f"{i+1}. SORTING SOLUTION: n= {n} Product: {lista[n - 1] * lista[n - 2]}")
  if i==99:
    print("Pass")

Writing max_pairwise_super_efficient_against_sorting_stress_tests.py


In [156]:
!python max_pairwise_super_efficient_against_sorting_stress_tests.py

1. SUPER EFFICENT SOLITION: n= 86 Product= 630550657
1. SORTING SOLUTION: n= 86 Product: 630550657
2. SUPER EFFICENT SOLITION: n= 74 Product= 638716064
2. SORTING SOLUTION: n= 74 Product: 638716064
3. SUPER EFFICENT SOLITION: n= 18 Product= 599686675
3. SORTING SOLUTION: n= 18 Product: 599686675
4. SUPER EFFICENT SOLITION: n= 4 Product= 401862930
4. SORTING SOLUTION: n= 4 Product: 401862930
5. SUPER EFFICENT SOLITION: n= 18 Product= 640039040
5. SORTING SOLUTION: n= 18 Product: 640039040
6. SUPER EFFICENT SOLITION: n= 12 Product= 526404592
6. SORTING SOLUTION: n= 12 Product: 526404592
7. SUPER EFFICENT SOLITION: n= 8 Product= 529542702
7. SORTING SOLUTION: n= 8 Product: 529542702
8. SUPER EFFICENT SOLITION: n= 12 Product= 496893830
8. SORTING SOLUTION: n= 12 Product: 496893830
9. SUPER EFFICENT SOLITION: n= 7 Product= 450194280
9. SORTING SOLUTION: n= 7 Product: 450194280
10. SUPER EFFICENT SOLITION: n= 37 Product= 630920897
10. SORTING SOLUTION: n= 37 Product: 630920897
11. SUPER EFFI